rqt_gui/src/rqt_gui/ros2_plugin_context.py

```diff
@@ -0,0 +1,142 @@
+# Copyright (c) 2018 PickNik Consulting
+# All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions
+# are met:
+#
+#   * Redistributions of source code must retain the above copyright
+#     notice, this list of conditions and the following disclaimer.
+#   * Redistributions in binary form must reproduce the above
+#     copyright notice, this list of conditions and the following
+#     disclaimer in the documentation and/or other materials provided
+#     with the distribution.
+#   * Neither the name of the TU Darmstadt nor the names of its
+#     contributors may be used to endorse or promote products derived
+#     from this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+# COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+
+from qt_gui.plugin_context import PluginContext
+
+
+class Ros2PluginContext(PluginContext):
+    """
+    Provides information to the plugin and exposes methods to interact with the framework.
+
+    PluginContext relays all methods to the corresponding `PluginHandler`.
+    """
+
+    def __init__(self, handler, node):
```

Comment:
See my comments on https://github.com/ros-visualization/rqt_top/pull/6 about these helper functions. It seems like it would be better to simply pass out a copy of the node... Any reason why we wouldn't want to do that. 

..I realize that I previously requested the get_topic_names_and_types method here but maybe that was a bad idea

Reply:
The concern is just that someone writes a plugin and calls a mutable method on node without realizing the implications of doing so. That will impact the other plugins running on their system.

There is no way to prevent this from happening 100%, but a wrapper would heavily discourage calling these mutable functions. I did leave in the create_xyz and destroy_xyz methods because it's fair to expect the plugin to keep track of creation and destruction of the publishers/subscribers/clients/services that they need.

Another option would be to create a copy and overwrite __class__ of the copied node with a subclass that unimplements the methods that should not be available to plugins.

If they want a node that they can directly access, the plugin can always create one themselves.

Reply:
I am not convinced this is necessary. The big disadvantage is that this needs to be updated whenever a new method is added. E.g. currently the getter for the clock is missing from that interface. On the other hand it can't prevent users from accessing undesired methods anyway since they can just get `_node`.

Therefore I suggest to just add `node` to the context. The same needs to happen for the C++ equivalent.

Reply:
Sounds good. Will revert it back